### DISCUSSION TITLE
Adding a caption to an animated gif

### DISCUSSION THREAD

**Discussion by TheVexedGerman** [ImageMagick/ImageMagick#6825] [Help] [answered]
### ImageMagick version

7.1.1-20

### Operating system, version and so on

Manjaro Linux Kernel 6.1.55-1

### Description

I'd like to resize and add captions to the top of images. Doing this for static images works well with
```bash
magick  -size 500x -background white -gravity center -pointsize 30 caption:"caption text"  in.png -resize 500x -append out.png
```
but for animated gifs it appends all frames creating a long static image which isn't what I want.

Getting some inspiration from older help discussions I've come up with:
```bash
magick \
  in.gif \
  -layers coalesce \
  -resize 500x -background None -gravity North -splice 0x32+0+0 \
  NULL: \
  -size 500x -background white -gravity center -pointsize 30 caption:"caption text" \
  -gravity North -compose Over -layers composite \
  -layers optimize \
  out.gif
```
But that overwrites part of the image if the caption is long enough to become multi line. Is there a way to dynamically expand the gif based on the generated caption height or append the caption to every frame like the first command does to static images?

**Comment by snibgo** [comment]
Before the `NULL:`, you could `-splice` the frames, adding any number of rows to the top of each frame, so the caption will occupy that area.

To save guesswork, you could start the command by making a `caption:` image. The height of that image is the number of rows you need to `-splice`. Then `+delete that image before reading `in.gif`.

**Reply by TheVexedGerman** [reply]
Maybe I'm misunderstanding you, but I am already splicing in an empty area, I just don't know how to automatically get the vertical height of the caption.

I'm also not sure how exactly I'm supposed to get the height from the temp caption you've suggested. I've tried
```bash
caption = "caption text"
magick \
  -size 500x -background white -gravity center -pointsize 30 caption:$caption +delete \
  in.gif  \
  -layers coalesce \
  -resize 500x -background None -gravity North -splice '0x%[fx:u.h]+0+0' \
  NULL: \
  -size 500x -background white -gravity center -pointsize 30 caption:$caption \
  -gravity North -compose Over -layers composite \
  -layers optimize \
  out.gif
```

based on your suggestion but that results in both only one frame of the gif being captioned, but also the height of the gif being used.
With a bit of parenthesis usage and reordering I get:

```bash
magick \
  -size 500x -background white -gravity center -pointsize 30 caption:$caption \
  \( in.gif  \
  -layers coalesce \
  -resize 500x \) -background None -gravity North -splice '0x%[fx:u.h]+0+0' -delete 0 \ 
  NULL: \
  -size 500x -background white -gravity center -pointsize 30 caption:$caption \
  -gravity North -compose Over -layers composite \
  -layers optimize \
  out.gif
```

which also doesn't quite work as expected with the spliced area being double what is required, but simply adding a /2 to the fx expression leads to part of the image being cropped. Seemingly the full height is too much but the half height is too little.

**Comment by snibgo** [comment, marked as answer]
You need to save the height of the first caption image before deleting it, then use that saved value in the splice. 

I suggest this bash script:
```
caption="caption text"

magick \
  -size 500x -background white -gravity center -pointsize 30 caption:"$caption" \
  -set option:HH %h \
  +delete \
  in.gif  \
  -layers coalesce \
  -resize 500x -background None -gravity North -splice '0x%[HH]+0+0' \
  NULL: \
  -size 500x -background white -gravity center -pointsize 30 caption:"$caption" \
  -gravity North -compose Over -layers composite \
  -layers optimize \
  out.gif
```

**Reply by TheVexedGerman** [reply]
Oooh, I was missing the -set option. Must have overlooked that flag. Thank you.

I do wonder why the height was double the intended one in my second solution but that is just curiosity.

Have a wonderful day.

**Comment by snibgo** [comment]
> I do wonder why the height was double the intended one in my second solution but that is just curiosity.

Your code added `u.h` to each of the images. `u.h` is the height of the first image. So the height of the fitst image is added to the first image, which doubles its height. Then the new height of the first image is added to each of the other images.
